token-lending/program/src/processor.rs

```diff
@@ -1696,6 +1700,88 @@ fn process_flash_loan(
+    let token_program_id = next_account_info(account_info_iter)?;
+
+    let lending_market = LendingMarket::unpack(&lending_market_info.data.borrow())?;
+    if lending_market_info.owner != program_id {
+        msg!("Lending market provided is not owned by the lending program");
+        return Err(LendingError::InvalidAccountOwner.into());    
+    }
+    if &lending_market.token_program_id != token_program_id.key {
+        msg!("Lending market token program does not match the token program provided");
+        return Err(LendingError::InvalidTokenProgram.into());    
+    }
+    if &lending_market.owner != lending_market_authority_info.key {
+        msg!("Lending market owner does not match the lending market owner provided");
+        return Err(LendingError::InvalidMarketOwner.into());
+    }
+
+    let reserve = Reserve::unpack(&reserve_info.data.borrow())?;
+    if &reserve.liquidity.supply_pubkey == destination_info.key {
+        msg!("Reserve liquidity supply cannot be used as the destination liquidity provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    let obligation = Obligation::unpack(&obligation_info.data.borrow())?;
+    if obligation_info.owner != program_id {
+        msg!("Obligation provided is not owned by the lending program");
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &obligation.lending_market != lending_market_info.key {
+        msg!("Obligation lending market does not match the lending market provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if &obligation.owner != obligation_owner_info.key {
+        msg!("Obligation owner does not match the obligation owner provided");
+        return Err(LendingError::InvalidObligationOwner.into());
+    }
+    if !&obligation.borrows.is_empty() {
+        msg!("Obligation borrows must be zero");
+        return Err(LendingError::NonZeroObligationBorrow.into());
+    }
+
+    if !obligation_owner_info.is_signer {
+        msg!("Obligation owner provided must be a signer");
+        return Err(LendingError::InvalidSigner.into());
+    }
+    
+    let authority_signer_seeds = &[
+        lending_market_info.key.as_ref(),
+        &[lending_market.bump_seed],
+    ];
+
+    let lending_market_authority_pubkey = Pubkey::create_program_address(authority_signer_seeds, program_id)?;
+    if &lending_market_authority_pubkey != lending_market_authority_info.key{
+        msg!(
+            "Derived lending market authority does not match the lending market authority provided"
+        );
+        return Err(LendingError::InvalidMarketAuthority.into());    
+    }
+
+    spl_close_account(TokenCloseAccountParams{
+        token_program: token_program_id.clone(),
+        authority: lending_market_authority_info.clone(),
+        source: obligation_owner_info.clone(),
+        destination: destination_info.clone(),
+        authority_signer_seeds,
+    })?;
```

Comment:
In this case, `obligation_owner_info` isn't an SPL token account, but rather the actual user interacting with the system, which shouldn't be closed. Instead, you'll want to zero-out the data in `obligation_info` and send the lamports over to destination, implementing similar logic to the token program: https://github.com/solana-labs/solana-program-library/blob/51ece68cec173197d84c19630e3207410db7f58d/token/program/src/processor.rs#L695-L702

Reply:
Hi Jon, thanks for the feedback. `process_close_obligation_account` implementation uses processor.rs's `process_close_account`. Should this implementation disregard its use?

Reply:
Hi Jon, thanks for the feedback. process_close_obligation_account implementation uses processor.rs's process_close_account. Should this implementation disregard its use and/or only make changes to the obligation information?

Reply:
Hi Jon, thanks for the feedback. process_close_obligation_account implementation uses processor.rs's process_close_account. Should this implementation disregard its use and/or only make changes to the obligation information?  What is the lamport source?

Reply:
Your desire to reuse code is great, but in this case, the runtime won't accept it unfortunately, since it'll look like the token program is trying to move lamports that belong to the lending program.  So you'll make changes to the obligation data yourself, and the `obligation_info` is the source for the lamports.

Reply:
Got It, I'll find a different method instead of using 'process_close_account`. The source of lamports was misunderstood. The `obligation_info` is the object in obligation.rs correct?

Reply:
Kind of -- `obligation_info` is the name of the variable that you chose for the `AccountInfo` representing the obligation in your processor function

Reply:
The change is made, and it uses the memset method as well, which zeros the respective data in an `obligation`.  I'll create tests by the end of this week.